rholang/src/main/scala/rholang/interpreter/RholangADT.scala

```diff
@@ -0,0 +1,92 @@
+// Normalization:
+// Normalization requires the evaluation of everything that doesn't go through
+// the tuplespace, so the top level of a normalized process must have everything
+
+package coop.rchain.interpreter
+
+case class Par(
+  sends: List[Send],
+  receives: List[Receive],
+  // selects: List[Select],
+  evals: List[Eval],
+  news: List[New],
+  expr: List[Expr]
+  // matches: List[Match]
+) {
+  // TODO: write helper methods to append an X and return a new par
+  // TODO: write helper method to get an empty par
+  def this() =
+    this(List(), List(), List(), List(), List())
+}
+
+object Par {
+  def apply() : Par = new Par()
+}
+
+sealed trait Channel
+case class Quote(p: Par) extends Channel
+case class ChanVar(cvar: Var) extends Channel
+
+// While we use vars in both positions, when producing the normalized
+// representation we need a discipline to track whether a var is a name or a
+// process.
+// These are DeBruijn levels
+sealed trait Var
+case class BoundVar(level: Int) extends Var
+// It is an error to use WildCard() in a non-binding position
+case class WildCard() extends Var
+// Variables that occur free in Par used as a pattern or ChanVar are binders.
+// For the purpose of comparing patterns, we count just like BoundVars.
+
+// In the DeBruijn level paper, they use negatives, but this is more clear.
+case class FreeVar(level: Int) extends Var
+
+// upon send, all free variables in data are substituted with their values.
+// also if a process is sent, it is auto-quoted.
+case class Send(chan: Channel, data: Channel)
+
+// [Par] is an n-arity Pattern.
+// It's an error for free Variable in the list of patterns to overlap.
```

Comment:
This comment is hard to parse as written.  I think you mean that it's an error for a single free variable to occur in more than one place in a pattern.

Reply:
When I get to this part in the normalizer, I'll write a test.